### PR TITLE
Added reset password dialog upon SQL Server expired password error

### LINKED ISSUE #605: Feature: Change password of expired login
- SQL Operations Studio Version: 0.25.4

As a Linux user I have a SQL login with assigned password. After the password expires I am unable to reset it via SQL Operations Studio. Any attempts to connect via sqlops or sqlcmd fail with error message that the password for the login expired.

Steps to Reproduce:
- Create a SQL login whose password has to be changed
`CREATE LOGIN [test] WITH PASSWORD=N'test', DEFAULT_DATABASE=[master], DEFAULT_LANGUAGE=[us_english], CHECK_EXPIRATION=ON, CHECK_POLICY=ON
GO`
- Login at SQL Operations studio

It would be extremely helpful to have a dialogue pop up to prompt the user for the new password. This is what SSMS does by default.

### PR DESCRIPTION
**Pull request by smartguest**
This PR adds in a simple reset dialog that pops up whenever a password expired error is encountered while connecting to a profile.

The user can press OK to change the password (and connect to the server and save the password immediately if enabled). 

If an error is returned from STS when attempting to change the password, users will need to try connecting again to bring the dialog again so they can enter a different password.

Redesigned password reset dialog:
Password Change Dialog updated

Removed unnecessary dialog in error:
Password Change Dialog password cannot be changed

First connect for an expired password (initial unexpected error happens on SSMS as well):
ExpiredPasswordDialogOpen

Error for an invalid password (error from STS):
ExpiredPasswordDialogError

Loading spinner and successful connect for a valid password:
ExpiredPasswordDialogSuccess

This PR fixes #605

Companion PRS: 

https://github.com/microsoft/sqlops-dataprotocolclient/pull/71

https://github.com/microsoft/sqltoolsservice/pull/1771

### CODE REVIEW

File: src/sql/azdata.d.ts [hunk 1]
```diff
@@ -559,6 +559,8 @@ declare module 'azdata' {
 
 		disconnect(connectionUri: string): Thenable<boolean>;
 
+		changePassword(connectionUri: string, connectionInfo: ConnectionInfo, newPassword: string): Thenable<PasswordChangeResult>;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.d.ts:562]: All API updates should go in `azdata.proposed.d.ts`

And also have doc comments added.

File: src/sql/platform/connection/common/constants.ts [hunk 2]
```diff
@@ -65,3 +65,6 @@ export const UNSAVED_GROUP_ID = 'unsaved';
 /* Server Type Constants */
 export const sqlDataWarehouse = 'Azure SQL Data Warehouse';
 export const gen3Version = 12;
+
+/* Password Reset Error code */
+export const passwordErrorCode = 18488;
```
Comment (Charles-Gagnon) [inline src/sql/platform/connection/common/constants.ts:70]: This is specific to MSSQL - but the feature itself seems generic to all providers. Is the intent for this to be something that all providers can get automatically? If so you're going to need to refactor this to have the provider be able to tell you when an error is specifically because a password change is needed.

Comment (smartguest) [reply src/sql/platform/connection/common/constants.ts:70]: For now the feature is only compatible with SqlServer as the change password function is part of the SqlClient library (seems to be meant for Sql Servers)

Comment (Charles-Gagnon) [reply src/sql/platform/connection/common/constants.ts:70]: Well, that's just an implementation detail. Providers would be the one actually doing the calls (which in the case of MSSQL is STS) - but other providers could theoretically want to hook into this and have their own implementation called when needed using whatever method of changing the password is available to them.

Comment (smartguest) [reply src/sql/platform/connection/common/constants.ts]: Yeah true, but I'm not aware how any other provider uses the change password feature. For now, I'm focused on Sql Server (though I will need to modify the provider class to have the password change needed indicator for another time).

**Review by erinstellato-ms — COMMENTED** [review]
With regard to this:

"If an error is returned from STS when attempting to change the password, users will need to try connecting again to bring the dialog again so they can enter a different password."

What is the scenario where this would occur?  My concern is that users will be confused (frustrated) it's not clear what has happened - they will they changed their password, but (if I'm understanding correctly) some other error got in the way.  In their minds, they will have to start all over and might not understand why.

**Comment by smartguest** [comment]
@erinstellato-ms it will occur if a user decides to enter a password that does not meet the password requirement or if the user decides to enter an empty string for the password. An error message will pop up explaining why the password failed to be changed (it is the two error messages posted above), then once the user presses OK, the error dialog disappears along with the change password dialog. This means the user will have to try connecting again in this case.

**Comment by erinstellato-ms** [comment]
> @erinstellato-ms
> 
> It will occur if a user decides to enter a password that does not meet the password requirement or if the user decides to enter an empty string for the password.
> 
> An error message will pop up explaining why the password failed to be changed (it is the two error messages posted above, and they clearly state what went wrong)
> 
> Once the user presses OK, the error dialog disappears along with the change password dialog. This means the user will have to try connecting again in this case.

Can you share an example of the error message?  I'd like to view the text if possible.  Thanks!

**Comment by Charles-Gagnon** [comment]
Wouldn't it be better to not hide the password change dialog until the change is successful or the user cancels? If an error occurs you can display that, but you probably should keep the dialog showing so the user doesn't have to go through the entire process again.

**Comment by Charles-Gagnon** [comment]
(also I would suggest removing the callstack from the error message displayed. That isn't really needed and would just confuse users)

**Comment by smartguest** [comment]
Yeah I can @Charles-Gagnon just need to add a catch on the promise.

**Comment by smartguest** [comment]
Added updated error messages in the STS side, (the "Changed Database" lines are part of the error message from SqlClient and are out of our control, but the first lines of the code tell the user the error reason directly.

**Comment by smartguest** [comment]
Error recovery steps have been added to the error messages.

**Comment by kburtram** [comment]
@smartguest the UX on the dialog seems like it could be improved a bit for basic things like padding, labels, alignment.  Could you please sync with @alanrenmsft to get some suggestions for this?

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import { ILogService } from 'vs/platform/log/common/log';
+import { IAdsTelemetryService } from 'sql/platform/telemetry/common/telemetry';
+import { IConnectionDialogService } from 'sql/workbench/services/connection/common/connectionDialogService';
+import { attachModalDialogStyler } from 'sql/workbench/common/styler';
+import { ILayoutService } from 'vs/platform/layout/browser/layoutService';
+import { ITextResourcePropertiesService } from 'vs/editor/common/services/textResourceConfiguration';
+import { LoadingSpinner } from 'sql/base/browser/ui/loadingSpinner/loadingSpinner';
+
+
+const OkText: string = localize('passwordChangeDialog.ok', "OK");
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:30]: Why do these variables have inconsistent casing?   Shouldn't they all use `camelCasing`?

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+const CancelText: string = localize('passwordChangeDialog.cancel', "Cancel");
+const DialogTitle: string = localize('passwordChangeDialog.title', "Change Password");
+const TitleIconClass: string = 'icon filterLabel';
+const newPasswordText: string = localize('passwordChangeDialog.newPassword', 'New password for SQL Server Login:');
+const confirmPasswordText: string = localize('passwordChangeDialog.confirmPassword', 'Confirm password:');
+const connectCheckboxText: string = localize('passwordChangeDialog.connectText', 'Connect?:');
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:36]: I don't think the label should have a `?` in the text.  Also, since this is for a checkbox that already has a label I don't think we need this label in the UI.

File: src/sql/platform/connection/common/connectionManagement.ts [hunk 5]
```diff
@@ -71,6 +71,7 @@ export interface IConnectionResult {
 	callStack: string;
 	errorHandled?: boolean;
 	connectionProfile?: IConnectionProfile;
+	uriForPasswordChange?: string;
```
Comment (kburtram) [inline src/sql/platform/connection/common/connectionManagement.ts:74]: should this be name something more generic such as `connectionUri` or is this a different URI?

File: src/sql/azdata.proposed.d.ts [hunk 6]
```diff
@@ -385,6 +385,29 @@ declare module 'azdata' {
 		title: string;
 	}
 
+	export interface ConnectionProvider extends DataProvider {
+		/**
+		 * Changes the password for a SQL Server connection profile.
```
Comment (kburtram) [inline src/sql/azdata.proposed.d.ts:390]: probably shouldn't refer to "SQL Server" since presumably other providers could implement this method too?

**Comment by alanrenmsft** [comment]
1. the dialog title should be left aligned
2. I think we only need the following fields on the dialog: `New Password` and `Confirm Password`. The UI is triggered by connect action, I don't think we need to ask the user whether they want to connect.
3. we should use the dialog's built-in loading indicator on the left side of the dialog footer, see the connection dialog's connecting indicator for example.
4. the width of the dialog should be reduced; size allocation should be: auto for label column and all other available for the input box column to accommodate the localized strings.

also, could you please attach the screenshot from SSMS for comparison and reference. @smartguest

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 7]
```diff
@@ -275,10 +276,13 @@ export class ConnectionDialogService implements IConnectionDialogService {
 			} else if (connectionResult && connectionResult.errorHandled) {
 				this._connectionDialog.resetConnection();
 				this._logService.debug(`ConnectionDialogService: Error handled and connection reset - Error: ${connectionResult.errorMessage}`);
-			} else {
+			} else if (connectionResult.errorCode !== Constants.sqlPasswordErrorCode) {
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:279]: it seems like we should test `connectionResult.errorCode === Constants.sqlPasswordErrorCode` and do the password change.  And leave the default case the same as before.  i.e. making password change the default doesn't make sense from readability to me.

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+		const confirmPasswordRow = DOM.append(body, DOM.$('tr'));
+		DOM.append(confirmPasswordRow, DOM.$('td')).innerText = confirmPasswordText;
+		this._confirmValueText = new InputBox(DOM.append(confirmPasswordRow, DOM.$('.confirm-text')), this.contextViewService, {});
+		this._confirmValueText.inputElement.type = 'password';
+		this._register(attachInputBoxStyler(this._confirmValueText, this._themeService));
+
+		const saveAndCloseCheckboxRow = DOM.append(body, DOM.$('tr'));
+		DOM.append(saveAndCloseCheckboxRow, DOM.$('td')).innerText = connectCheckboxText;
+		this._connectOnClose = new Checkbox(DOM.append(saveAndCloseCheckboxRow, DOM.$('.connect-check')), { label: connectCheckboxLabel });
```
Comment (kburtram) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:113]: Do we need a checkbox for connect on close?  If user gets the error during connection shouldn't assume they want to connect (at least making the checkbox on by default).  Does SSMS have this checkbox in this scenario?

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts (hunk 3 above)
Comment (smartguest) [reply src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:30]: This is a holdover from the other dialogs, but I can change it here.

**Comment by smartguest** [comment]
@alanrenmsft @cheenamalhotra specifically asked me to include the ability to checkmark if we can connect or not, so I'm keeping that for now.

**Comment by alanrenmsft** [comment]
> @alanrenmsft @kburtram, @cheenamalhotra specifically asked me to include the ability to checkmark if we can connect or not, since this may be reused for right click to change password feature but yes this can be defaulted to on.

@cheenamalhotra any reason?

File: src/sql/azdata.proposed.d.ts [hunk 9]
```diff
@@ -385,6 +385,29 @@ declare module 'azdata' {
...
+	// Password Change Request ----------------------------------------------------------------------
+	export interface PasswordChangeResult {
+		/**
+		 * Whether the password change was successful
+		 */
+		result: boolean;
+		/**
+		 * The error message if the password change was unsuccessful
+		 */
+		messages?: string | undefined;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:404]: This seems like it should be `errorMessage` and then the one below `errorDetails` or something given the comments

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts (hunk 7 above)
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:279]: We should only do this for MSSQL providers - if another provider happened to give an error code with the same number that isn't necessarily going to mean that a password change is required (not to mention currently there doesn't seem to be any sort of generic way for providers to actually hook into this function as I discussed before. It's specific to just MSSQL)

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts (hunk 8 above)
Comment (cheenamalhotra) [reply src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:113]: I think this was added to allow users to not have to enter password again, once they reset and connect directly on close - but we can choose either way!

Comment (cheenamalhotra) [reply src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:113]: This is the behavior with change password in AAD which is a good reference point too...

image

Once user sets a new password, they simply "Sign in" and are logged in.

Comment (kburtram) [reply src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:113]: Personally I don't think we need the "Connect" checkbox (similar to SSMS).  Since the user already clicked Connect prior to getting the prompt we can infer their intent.  If the changed their mind or don't want to change password they can cancel the dialog.

**Comment by alanrenmsft** [comment]
@smartguest please update the screenshots/gifs after making changes.

**Comment by erinstellato-ms** [comment]
@smartguest I'm not sure how the dialog looks at this point, but I would agree that the Connect? checkbox is not needed.  After the user changes the password, I would expect the flow to be that they select OK, the password is updated, and assuming success, we try to login.

If you can post updated screenshots of the dialog, I'm happy to review those for spacing, color, etc.  Thanks!

File: src/sql/azdata.proposed.d.ts [hunk 10]
```diff
@@ -403,6 +403,25 @@ declare module 'azdata' {
 		title: string;
 	}
 
+	export interface ConnectionProvider extends DataProvider {
+		/**
+		 * Changes the password for a connection profile (in case of the password expiration scenario)
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:408]: I don't think we need to specify that this is for password expiration. After all -there's nothing stopping someone from calling changePassword directly on a provider (by calling azdata.dataprotocol.getProvider)

```suggestion
		 * Changes the password for a connection profile
```

File: src/sql/azdata.proposed.d.ts [hunk 11]
```diff
@@ -403,6 +403,25 @@ declare module 'azdata' {
...
+		 */
+		changePassword(connectionUri: string, connectionInfo: ConnectionInfo, newPassword: string): Thenable<PasswordChangeResult>;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:410]: This should be optional - we can't guarantee that existing providers will implement it and this isn't something that currently needs to be required (especially since it's currently only for MSSQL...)

```suggestion
		changePassword?(connectionUri: string, connectionInfo: ConnectionInfo, newPassword: string): Thenable<PasswordChangeResult>;
```

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 12]
```diff
@@ -288,6 +292,31 @@ export class ConnectionDialogService implements IConnectionDialogService {
...
 
+	public async changePasswordFunction(connection: IConnectionProfile, params: INewConnectionParams, uri: string, oldPassword: string, newPassword: string): Promise<void> {
+		if (oldPassword !== newPassword) {
+			this.showErrorDialog(Severity.Error, Constants.sqlPasswordChangeErrorHeader, Constants.sqlPasswordMismatchMessage);
+			return Promise.reject(new Error(Constants.sqlPasswordChangeErrorHeader));
+		}
+		let passwordChangeResult = await this._connectionManagementService.sendChangePassword(connection, uri, newPassword);
+		if (!passwordChangeResult.result) {
+			let detailMessage = passwordChangeResult.errorMessage;
+			if (detailMessage.indexOf(Constants.sqlPasswordDNMReqs) !== -1) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:303]: Why would you not just include these additional message details in the original error coming from STS? ADS shouldn't need to know anything about the specific details of the error or how to fix it.

**Review by Charles-Gagnon — COMMENTED** [review]
@kburtram @alanrenmsft @smartguest Is the intention for this to be generic functionality that all providers can hook in to? Right now it isn't really made that way - and if you only intend to use this for MSSQL then I'm not sure this should be included in the common ConnectionProvider API. We can just have it as an internal thing to start with until we take the time to actually make this extensible for all providers.

**Comment by smartguest** [comment]
@Charles-Gagnon should I still leave the actual implementation of change password in the Connection Service of STS, it still relies on code that is part of the implementation of the Connection Service for one of its checks, not to mention the handling of requests.

**Comment by Charles-Gagnon** [comment]
To be clear - what you have isn't going to "break" anything as far as I'm aware. My concern here is that you're adding stuff to the public extensibility layer that other extensions won't be able to use (since this is currently only done for MSSQL). 

You don't have to add things to the public API to do what you're trying to accomplish - we can have internal functionality that makes the RPC calls to the MSSQL extension directly to call changePassword. This of course isn't super great - ideally in my mind this functionality is something that all extensions could hook in to and utilize. But designing such a feature seems to be out of scope of what you're trying to accomplish so my suggestion is purely just to move things around so that we aren't confusing extension authors by providing a public API that does nothing for them.

**Comment by smartguest** [comment]
Added updated GIFs after changes

**Comment by cheenamalhotra** [comment]
I think 'Change Password' on server request/error number is not an uncommon scenario and can be extended beyond MSSQL provider. e.g. MySQL supports [Password expiration](https://dev.mysql.com/doc/refman/8.0/en/password-management.html#:~:text=MySQL%20supports%20these%20password-management%20capabilities%3A%201%20Password%20expiration%2C,a%20%20primary%20or%20secondary%20password.%20More%20items) which is something MySQL provider can make use of in future.

**Comment by Charles-Gagnon** [comment]
Definitely, I wasn't saying that it wouldn't be useful. Just that it currently isn't doing that.

And if you're looking to make this extensible I would very much suggest looking at a much more generic way of doing this - for example having a generic way for providers to register a callback that is called when an error occurs during a connection. This would allow them to do whatever they needed without us needing to cover every single possible scenario in core ourselves. 

Example flow : 

1. Provider is registered with an "onConnectionError" callback
2. User attempts a connection through that provider
3. Core calls the provider extension to make the connection
4. Provider connection call returns an error
5. Core calls the provider onConnectionError callback (if one exists, otherwise falls back to default of just showing error)
6. Provider does whatever it wants, such as showing a dialog or automatically trying to fix the issue and then returns the result (handled or not)
7. If result is handled core then re-attempts connection, otherwise it shows the error

We could then still provide simple ways for users to accomplish common tasks - such as showing the firewall rule dialog or change password dialog (so that each extension doesn't have to re-implement it themselves). This could be done either through the extension API or by a package we provide that they can reference separately.

File: src/sql/azdata.proposed.d.ts [hunk 13]
```diff
@@ -403,6 +403,25 @@ declare module 'azdata' {
 		title: string;
 	}
 
+	export interface ConnectionProvider extends DataProvider {
+		/**
+		 * Changes the password for a connection profile (for Azure Data Studio use only)
+		 */
+		changePassword?(connectionUri: string, connectionInfo: ConnectionInfo, newPassword: string): Thenable<PasswordChangeResult>;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts:410]: The comment is not accurate, it is not changing the password for a connection profile, this is only for the scenario of password expiration for SQL Authentication.

File: src/sql/platform/connection/common/connectionManagement.ts [hunk 14]
```diff
@@ -119,6 +120,11 @@ export interface IConnectionManagementService {
 	 */
 	connectAndSaveProfile(connection: IConnectionProfile, uri: string, options?: IConnectionCompletionOptions, callbacks?: IConnectionCallbacks): Promise<IConnectionResult>;
 
+	/**
+	 * Changes password for a SQL Server connection profile.
+	 */
+	sendChangePassword(connection: IConnectionProfile, uri: string, newPassword: string): Promise<azdata.PasswordChangeResult>;
```
Comment (alanrenmsft) [inline src/sql/platform/connection/common/connectionManagement.ts:126]: `changePassword`

File: src/sql/platform/connection/common/connectionManagement.ts [hunk 15]
```diff
@@ -71,6 +71,7 @@ export interface IConnectionResult {
 	callStack: string;
 	errorHandled?: boolean;
 	connectionProfile?: IConnectionProfile;
+	connectionUri?: string;
```
Comment (alanrenmsft) [inline src/sql/platform/connection/common/connectionManagement.ts:74]: why is this needed? connectionProfile property should include the connectionUri, right?

File: src/sql/platform/connection/common/constants.ts [hunk 16]
```diff
@@ -65,3 +65,8 @@ export const UNSAVED_GROUP_ID = 'unsaved';
 /* Server Type Constants */
 export const sqlDataWarehouse = 'Azure SQL Data Warehouse';
 export const gen3Version = 12;
+
+/* SQL Server Password Reset Constants */
+export const sqlPasswordErrorCode = 18488;
+export const sqlPasswordChangeErrorHeader = 'Failure when attempting to change password'
+export const sqlPasswordMismatchMessage = 'Passwords entered do not match\n\nPress OK and enter the exact same password in both boxes.';
```
Comment (alanrenmsft) [inline src/sql/platform/connection/common/constants.ts:72]: 1. we shouldn't have SQL Server specific stuff in core
2. shouldn't use English strings directly here.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 17]
```diff
@@ -288,6 +292,20 @@ export class ConnectionDialogService implements IConnectionDialogService {
 		}
 	}
 
+	public async changePasswordFunction(connection: IConnectionProfile, params: INewConnectionParams, uri: string, oldPassword: string, newPassword: string): Promise<void> {
```
Comment (alanrenmsft) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:295]: this should be in the password reset dialog

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+import { IAdsTelemetryService } from 'sql/platform/telemetry/common/telemetry';
+import { IConnectionDialogService } from 'sql/workbench/services/connection/common/connectionDialogService';
+import { attachModalDialogStyler } from 'sql/workbench/common/styler';
+import { ILayoutService } from 'vs/platform/layout/browser/layoutService';
+import { ITextResourcePropertiesService } from 'vs/editor/common/services/textResourceConfiguration';
+
+
+const okText: string = localize('passwordChangeDialog.ok', "OK");
+const cancelText: string = localize('passwordChangeDialog.cancel', "Cancel");
+const dialogTitle: string = localize('passwordChangeDialog.title', "Change your expired password");
```
Comment (alanrenmsft) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:30]: Change Password

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+const newPasswordText: string = localize('passwordChangeDialog.newPassword', 'New password for SQL Server Login:');
```
Comment (alanrenmsft) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:31]: no `SQL Server`, simply use `New password`

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+		this.title = dialogTitle;
+		this._register(attachModalDialogStyler(this, this._themeService));
+		this._okButton = this.addFooterButton(okText, () => this.handleOkButtonClick());
+		this._cancelButton = this.addFooterButton(cancelText, () => this.hide('cancel'), 'right', true);
+		this._register(attachButtonStyler(this._okButton, this._themeService));
+		this._register(attachButtonStyler(this._cancelButton, this._themeService));
+	}
+
+	protected renderBody(container: HTMLElement) {
+		const body = DOM.append(container, DOM.$('.change-password-dialog'));
```
Comment (alanrenmsft) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:83]: Please refer to my earlier comments about the space allocation.

**Review by alanrenmsft — CHANGES_REQUESTED** [review]

**Comment by alanrenmsft** [comment]
@smartguest let's talk about the feedback in our meeting today.

File: src/sql/platform/connection/common/connectionManagement.ts (hunk 15 above)
Comment (smartguest) [reply src/sql/platform/connection/common/connectionManagement.ts:74]: As I recall, when connecting for the first time, the connectionUri is blank, so if we never connect at all for the first time, we wouldn't have the necessary connectionUri to make the password change. I will have to verify this is the case.

Comment (smartguest) [reply src/sql/platform/connection/common/connectionManagement.ts:74]: The connection profile does not include the URI unfortunately, but we can generate it using the utils

**Comment by cheenamalhotra** [comment]
> Definitely, I wasn't saying that it wouldn't be useful. Just that it currently isn't doing that (we specifically check for MSSQL and only our specific error code).
> 
> And if you're looking to make this extensible I would very much suggest looking at a much more generic way of doing this - for example having a generic way for providers to register a callback that is called when an error occurs during a connection. This would allow them to do whatever they needed without us needing to cover every single possible scenario in core ourselves.
> 
> Example flow :
> 
> 1. Provider is registered with an "onConnectionError" callback
> 2. User attempts a connection through that provider
> 3. Core calls the provider extension to make the connection
> 4. Provider connection call returns an error
> 5. Core calls the provider onConnectionError callback (if one exists, otherwise falls back to default of just showing error)
> 6. Provider does whatever it wants, such as showing a dialog or automatically trying to fix the issue and then returns the result (handled or not)
> 7. If result is handled core then re-attempts connection, otherwise it shows the error
> 
> We could then still provide simple ways for users to accomplish common tasks - such as showing the firewall rule dialog or change password dialog (so that each extension doesn't have to re-implement it themselves). This could be done either through the extension API or by a package we provide that they can reference separately.

@Charles-Gagnon as discussed offline, we would take this feedback as a follow-up investigation activity as its the initial design limitation and not specific to the proposed changes. It would be ideal to address it altogether with other error codes too.

**Comment by erinstellato-ms** [comment]
@smartguest I have a question about this comment in the updated screenshots (thank you for those): "First connect for an expired password (initial unexpected error happens on SSMS as well)" Do we know why that fails and can we handle it better?

**Review by erinstellato-ms — COMMENTED** [review]
For the password cannot be changed error...  Do we have to include the "Changed database context to 'master'" and "Changed language setting to us_english" text?  It doesn't seem relevant to the error, and is not helpful for users.

File: src/sql/workbench/services/connection/browser/passwordChangeDialog.ts [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+		@IContextViewService private readonly contextViewService: IContextViewService,
+		@ITextResourcePropertiesService textResourcePropertiesService: ITextResourcePropertiesService,
+	) {
+		super('', '', telemetryService, layoutService, clipboardService, themeService, logService, textResourcePropertiesService, contextKeyService, { hasSpinner: true, spinnerTitle: passwordChangeLoadText, dialogStyle: 'normal', width: dialogWidth, dialogPosition: 'left' });
+	}
+
+	public open(profile: IConnectionProfile, params: INewConnectionParams) {
+		this._profile = profile;
+		this._params = params;
+		//this._connectionService = connectionService;
```
Comment (cheenamalhotra) [inline src/sql/workbench/services/connection/browser/passwordChangeDialog.ts:69]: Remove commented code

File: src/sql/platform/connection/common/constants.ts [hunk 22]
```diff
@@ -65,3 +65,6 @@ export const UNSAVED_GROUP_ID = 'unsaved';
 /* Server Type Constants */
 export const sqlDataWarehouse = 'Azure SQL Data Warehouse';
 export const gen3Version = 12;
+
+/* SQL Server Password Reset Constants */
```
Comment (cheenamalhotra) [inline src/sql/platform/connection/common/constants.ts:69]:
```suggestion
/* SQL Server Password Reset Error code */
```

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts [hunk 23]
```diff
@@ -495,6 +504,12 @@ export class ConnectionDialogService implements IConnectionDialogService {
 		recentConnections.forEach(conn => conn.dispose());
 	}
 
+	public launchChangePasswordDialog(profile: IConnectionProfile, params: INewConnectionParams): void {
```
Comment (cheenamalhotra) [inline src/sql/workbench/services/connection/browser/connectionDialogService.ts:507]: Please add doc for public API

**Review by cheenamalhotra — APPROVED** [review]
Other than some minor changes, looks good.

**Comment by smartguest** [comment]
@erinstellato-ms The first connect error happens when you attempt to connect to a server using credentials for the first time, the exact error is: 

"A connection was successfully established with the server, but then an error occurred during the login process. (provider: Shared Memory Provider, error: 0 - No process is on the other end of the pipe.)"

This will have to be a separate issue to be looked into as it also happens with SSMS so it's not really part of this PR.

Also those messages are part of the SqlClient error message as it is and processing them individually is not an ideal solution, but we can try to filter them from the string.

File: src/sql/workbench/services/connection/browser/connectionDialogService.ts (hunk 23 above)
Comment (smartguest) [reply src/sql/workbench/services/connection/browser/connectionDialogService.ts:507]: Done so in the IConnectionDialogService interface definition (located in src/sql/workbench/services/connection/common/connectionDialogService.ts), but I'll add the comment here as well

**Comment by Charles-Gagnon** [comment]
@smartguest This PR depends on both the STS changes and (to a slightly lesser extent) the dataprocol-client changes. 

You typically should be including all required changes together in a single PR so we don't have broken features in ADS - which normally means waiting until a build of STS is published that contain the required changes. (and a new release of dataprotocol-client should be done as well)

**Comment by sirio3mil** [comment]
It is merged but has it already been included at least in the insider version? Yesterday it expired and I did not get the pop up although it is possible that I did not have the latest build, now I have version 1.42.0-insider, yesterday I may have had the previous one

**Comment by rick-syncbak** [comment]
I'm still seeing this in version 1.48.0 in 2024. Every time I create a new user who uses Azure Data Studio with "User must change password at net login" , we have to have them use a desktop w/ SSMS to change their password.

**Comment by erinstellato** [comment]
@smartguest can you re-test this in ADS 1.48 when you have a minute?  Thanks.  

@kisantia for awareness.

**Comment by smartguest** [comment]
@erinstellato @rick-syncbak It appears the current code only handles one kind of password expired error: 18488, which is the Password needs to be changed (and not Password has expired which is 18487), I will add support for that number as an alternative check: https://learn.microsoft.com/en-us/sql/relational-databases/native-client/features/changing-passwords-programmatically?view=sql-server-ver15

**Comment by smartguest** [comment]
@erinstellato It works for error 18488 (The one that states password needs to be changed and it should work for error 18487, which is the password expired error as addressed in this PR): https://github.com/microsoft/azuredatastudio/issues/22916). 

Can you get the exact error from the "Copy Details" button when it happens? @rick-syncbak and paste it here? I need to identify to see if the error you are encountering matches the error code that we currently check for).

**Comment by smartguest** [comment]
Here is a GIF of the 1.48 showing me creating the login on a non cloud SQL Server Instance with SQL Login, with password expired for next login, and then logging in with successful results:

PasswordExpired

**Comment by rick-syncbak** [comment]
I was able to test successfully on a "must change password on first login"
scenario.

Thank you!

**Comment by erinstellato-ms** [comment]
@rick-syncbak Thanks for following up, glad it's working for you!

Thanks @smartguest for validating and sharing the video!
